### ISSUE TITLE
Attention masking in TSDAE's DenoisingAutoEncoderLoss

### ISSUE THREAD

**Issue by ghadialhajj** [UKPLab/sentence-transformers#3268]
In the decoder of the DenoisingAutoEncoderLoss class, no attention mask is passed and thus, if I understand correctly, padding tokens do take part in the attention weights calculation still. 

It is true that padding tokens are ignored during loss calculation (and that the self attention applies causal masking, but that doesn't address attending to padding tokens in particular). However, wouldn't this still mean that padding tokens would still take part in attention and thus mess up with the attention weights through the softmax?

Could someone confirm or refute my understanding?

**Comment by HuangBugWei** [comment]
Hi @ghadialhajj, I was also confused when I saw the warning: 
>We strongly recommend passing in an `attention_mask` since your input_ids may be padded. See https://huggingface.co/docs/transformers/troubleshooting#incorrect-output-when-padding-tokens-arent-masked.

After looking into it a bit, to my best knowledge, because the softmax is applied row-wise in the attention mechanism, padding tokens that appear after the ending token should not negatively impact the attention weights for the preceding valid tokens. Furthermore, in this specific task, padding is always placed after the end token (to make the batch data have the same length), and the loss function explicitly ignores padding tokens: ```ce_loss_fct = nn.CrossEntropyLoss(ignore_index=self.tokenizer_decoder.pad_token_id).```

Therefore, I believe it is acceptable not to pass an explicit attention mask during the decoder part in this particular scenario. My simple illustration below further clarifies this point."

Below is my simple illustration.

![Image](https://github.com/user-attachments/assets/92b02480-ee9a-4ce8-9cf4-a8e2fe17c03c)

**Comment by tomaarsen** [comment]
Hello!

Apologies for the delay. I've also looked into this, and I experimented with a version of `DenoisingAutoEncoderLoss` that uses:
```python
        decoder_outputs_attn = self.decoder(
            input_ids=decoder_input_ids,
            inputs_embeds=None,
            attention_mask=target_features["attention_mask"][:, : target_length - 1],  # <- This is new!
            encoder_hidden_states=reps[:, None],  # (bsz, hdim) -> (bsz, 1, hdim)
            encoder_attention_mask=source_features["attention_mask"][:, 0:1],
            labels=None,
            return_dict=None,
            use_cache=False,
        )
```
And I'm noticing that the losses from each of these are exactly identical. If I try both approaches in the same loss function back to back, then the original implementation is approximately 2.5% faster on average across the first 1000 steps of a training run that I tried, if that original implementation is ran *first*. If it is ran *second*, then it is approximately 11.7% faster on average across the first 1000 steps.

Here's my updated `DenoisingAutoEncoderLoss` to try and reproduce it:

<details><summary>DenoisingAutoEncoderLoss</summary>

```python
from __future__ import annotations

import logging
from collections.abc import Iterable
import time

from torch import Tensor, nn
from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer, PreTrainedModel

from sentence_transformers import SentenceTransformer
from sentence_transformers.models import StaticEmbedding

logger = logging.getLogger(__name__)


class DenoisingAutoEncoderLoss(nn.Module):
    def __init__(
        self, model: SentenceTransformer, decoder_name_or_path: str | None = None, tie_encoder_decoder: bool = True
    ) -> None:
        r"""
        This loss expects as input a pairs of damaged sentences and the corresponding original ones.
        During training, the decoder reconstructs the original sentences from the encoded sentence embeddings.
        Here the argument 'decoder_name_or_path' indicates the pretrained model (supported by Hugging Face) to be used as the decoder.
        Since decoding process is included, here the decoder should have a class called XXXLMHead (in the context of Hugging Face's Transformers).
        The 'tie_encoder_decoder' flag indicates whether to tie the trainable parameters of encoder and decoder,
        which is shown beneficial to model performance while limiting the amount of required memory.
        Only when the encoder and decoder are from the same architecture, can the flag 'tie_encoder_decoder' work.

        The data generation process (i.e. the 'damaging' process) has already been implemented in ``DenoisingAutoEncoderDataset``,
        allowing you to only provide regular sentences.

        Args:
            model (SentenceTransformer): The SentenceTransformer model.
            decoder_name_or_path (str, optional): Model name or path for initializing a decoder (compatible with Hugging Face's Transformers). Defaults to None.
            tie_encoder_decoder (bool): Whether to tie the trainable parameters of encoder and decoder. Defaults to True.

        References:
            * TSDAE paper: https://arxiv.org/pdf/2104.06979.pdf
            * `Unsupervised Learning > TSDAE <../../examples/unsupervised_learning/TSDAE/README.html>`_

        Requirements:
            1. The decoder should have a class called XXXLMHead (in the context of Hugging Face's Transformers)
            2. Should use a large corpus

        Inputs:
            +------------------------------------------------------+--------+
            | Texts                                                | Labels |
            +======================================================+========+
            | (damaged\_sentence, original\_sentence) pairs        | none   |
            +------------------------------------------------------+--------+
            | sentence fed through ``DenoisingAutoEncoderDataset`` | none   |
            +------------------------------------------------------+--------+

        Example:
            ::

                from sentence_transformers import SentenceTransformer, losses
                from sentence_transformers.datasets import DenoisingAutoEncoderDataset
                from torch.utils.data import DataLoader

                model_name = "bert-base-cased"
                model = SentenceTransformer(model_name)
                train_sentences = [
                    "First training sentence", "Second training sentence", "Third training sentence", "Fourth training sentence",
                ]
                batch_size = 2
                train_dataset = DenoisingAutoEncoderDataset(train_sentences)
                train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True, drop_last=True)
                train_loss = losses.DenoisingAutoEncoderLoss(
                    model, decoder_name_or_path=model_name, tie_encoder_decoder=True
                )
                model.fit(
                    train_objectives=[(train_dataloader, train_loss)],
                    epochs=10,
                )
        """
        super().__init__()

        if isinstance(model[0], StaticEmbedding):
            raise ValueError(
                "DenoisingAutoEncoderLoss is not compatible with a SentenceTransformer model based on a StaticEmbedding."
            )

        self.encoder = model  # This will be the final model used during the inference time.
        self.tokenizer_encoder = model.tokenizer

        encoder_name_or_path = model[0].auto_model.config._name_or_path
        if decoder_name_or_path is None:
            assert (
                tie_encoder_decoder
            ), "Must indicate the decoder_name_or_path argument when tie_encoder_decoder=False!"
        if tie_encoder_decoder:
            if decoder_name_or_path:
                logger.warning("When tie_encoder_decoder=True, the decoder_name_or_path will be invalid.")
            decoder_name_or_path = encoder_name_or_path

        self.tokenizer_decoder = AutoTokenizer.from_pretrained(decoder_name_or_path)
        self.need_retokenization = not isinstance(self.tokenizer_encoder, type(self.tokenizer_decoder))

        decoder_config = AutoConfig.from_pretrained(decoder_name_or_path)
        decoder_config.is_decoder = True
        decoder_config.add_cross_attention = True
        kwargs_decoder = {"config": decoder_config}
        try:
            self.decoder = AutoModelForCausalLM.from_pretrained(decoder_name_or_path, **kwargs_decoder)
        except ValueError as e:
            logger.error(
                f'Model name or path "{decoder_name_or_path}" does not support being as a decoder. Please make sure the decoder model has an "XXXLMHead" class.'
            )
            raise e
        assert model[0].auto_model.config.hidden_size == decoder_config.hidden_size, "Hidden sizes do not match!"
        if self.tokenizer_decoder.pad_token is None:
            # Needed by GPT-2, etc.
            self.tokenizer_decoder.pad_token = self.tokenizer_decoder.eos_token
            self.decoder.config.pad_token_id = self.decoder.config.eos_token_id

        if len(AutoTokenizer.from_pretrained(encoder_name_or_path)) != len(self.tokenizer_encoder):
            logger.warning(
                "WARNING: The vocabulary of the encoder has been changed. One might need to change the decoder vocabulary, too."
            )

        if tie_encoder_decoder:
            assert not self.need_retokenization, "The tokenizers should be the same when tie_encoder_decoder=True."
            if len(self.tokenizer_encoder) != len(self.tokenizer_decoder):  # The vocabulary has been changed.
                self.tokenizer_decoder = self.tokenizer_encoder
                self.decoder.resize_token_embeddings(len(self.tokenizer_decoder))
                logger.warning(
                    "Since the encoder vocabulary has been changed and --tie_encoder_decoder=True, now the new vocabulary has also been used for the decoder."
                )
            decoder_base_model_prefix = self.decoder.base_model_prefix
            try:
                # Compatibility with transformers <4.40.0
                PreTrainedModel._tie_encoder_decoder_weights(
                    model[0].auto_model,
                    self.decoder._modules[decoder_base_model_prefix],
                    self.decoder.base_model_prefix,
                )
            except TypeError:
                # Compatibility with transformers >=4.40.0
                PreTrainedModel._tie_encoder_decoder_weights(
                    model[0].auto_model,
                    self.decoder._modules[decoder_base_model_prefix],
                    self.decoder.base_model_prefix,
                    encoder_name_or_path,
                )
        self.ratios = []

    def retokenize(self, sentence_features: dict[str, Tensor]) -> dict[str, Tensor]:
        input_ids = sentence_features["input_ids"]
        device = input_ids.device
        sentences_decoded = self.tokenizer_encoder.batch_decode(
            input_ids, skip_special_tokens=True, clean_up_tokenization_spaces=True
        )
        retokenized = self.tokenizer_decoder(
            sentences_decoded, padding=True, truncation="longest_first", return_tensors="pt", max_length=None
        ).to(device)
        return retokenized

    def forward(self, sentence_features: Iterable[dict[str, Tensor]], labels: Tensor) -> Tensor:
        source_features, target_features = tuple(sentence_features)
        if self.need_retokenization:
            # since the sentence_features here are all tokenized by encoder's tokenizer,
            # retokenization by the decoder's one is needed if different tokenizers used
            target_features = self.retokenize(target_features)
        reps = self.encoder(source_features)["sentence_embedding"]  # (bsz, hdim)

        # Prepare input and output
        target_length = target_features["input_ids"].shape[1]
        decoder_input_ids = target_features["input_ids"].clone()[:, : target_length - 1]
        label_ids = target_features["input_ids"][:, 1:]

        # Decode
        attn_start_time = time.time()
        decoder_outputs_attn = self.decoder(
            input_ids=decoder_input_ids,
            inputs_embeds=None,
            attention_mask=target_features["attention_mask"][:, : target_length - 1],
            encoder_hidden_states=reps[:, None],  # (bsz, hdim) -> (bsz, 1, hdim)
            encoder_attention_mask=source_features["attention_mask"][:, 0:1],
            labels=None,
            return_dict=None,
            use_cache=False,
        )
        attn_time_delta = time.time() - attn_start_time
        og_start_time = time.time()
        decoder_outputs_og = self.decoder(
            input_ids=decoder_input_ids,
            inputs_embeds=None,
            attention_mask=None,
            encoder_hidden_states=reps[:, None],  # (bsz, hdim) -> (bsz, 1, hdim)
            encoder_attention_mask=source_features["attention_mask"][:, 0:1],
            labels=None,
            return_dict=None,
            use_cache=False,
        )
        og_time_delta = time.time() - og_start_time
        ratio = attn_time_delta / og_time_delta
        self.ratios.append(ratio)
        logger.info(f"Timings: {og_time_delta:.5f} (w/o attn mask) vs {attn_time_delta:.5f} (w/ attn mask), ratio: {ratio:.5f}, running ratio: {sum(self.ratios) / len(self.ratios):.5f}")

        # Calculate loss
        lm_logits = decoder_outputs_og[0]
        ce_loss_fct = nn.CrossEntropyLoss(ignore_index=self.tokenizer_decoder.pad_token_id)
        loss = ce_loss_fct(lm_logits.view(-1, lm_logits.shape[-1]), label_ids.reshape(-1))
        lm_logits = decoder_outputs_attn[0]
        ce_loss_fct = nn.CrossEntropyLoss(ignore_index=self.tokenizer_decoder.pad_token_id)
        loss_attn = ce_loss_fct(lm_logits.view(-1, lm_logits.shape[-1]), label_ids.reshape(-1))
        assert loss == loss_attn, "The loss should be the same when using attention mask or not."
        return loss

    @property
    def citation(self) -> str:
        return """
@inproceedings{wang-2021-TSDAE,
    title = "TSDAE: Using Transformer-based Sequential Denoising Auto-Encoderfor Unsupervised Sentence Embedding Learning",
    author = "Wang, Kexin and Reimers, Nils and Gurevych, Iryna",
    booktitle = "Findings of the Association for Computational Linguistics: EMNLP 2021",
    month = nov,
    year = "2021",
    address = "Punta Cana, Dominican Republic",
    publisher = "Association for Computational Linguistics",
    pages = "671--688",
    url = "https://arxiv.org/abs/2104.06979",
}
"""
```

</details>

E.g. a part of the outputs:
```
  5%|██████                                                                                                                    | 994/20055 [02:21<47:21,  6.71it/s]
2025-03-19 15:03:58 - Timings: 0.01703 (w/o attn mask) vs 0.01959 (w/ attn mask), ratio: 1.15016, running ratio: 1.11739
  5%|██████                                                                                                                    | 995/20055 [02:21<47:42,  6.66it/s]
2025-03-19 15:03:58 - Timings: 0.02420 (w/o attn mask) vs 0.02681 (w/ attn mask), ratio: 1.10772, running ratio: 1.11738
  5%|██████                                                                                                                    | 996/20055 [02:21<48:22,  6.57it/s]
2025-03-19 15:03:59 - Timings: 0.01613 (w/o attn mask) vs 0.01801 (w/ attn mask), ratio: 1.11666, running ratio: 1.11738
  5%|██████                                                                                                                    | 997/20055 [02:21<47:11,  6.73it/s]
2025-03-19 15:03:59 - Timings: 0.01708 (w/o attn mask) vs 0.01851 (w/ attn mask), ratio: 1.08404, running ratio: 1.11735
  5%|██████                                                                                                                    | 998/20055 [02:21<46:18,  6.86it/s]
2025-03-19 15:03:59 - Timings: 0.01894 (w/o attn mask) vs 0.02169 (w/ attn mask), ratio: 1.14512, running ratio: 1.11738
  5%|██████                                                                                                                    | 999/20055 [02:21<47:30,  6.68it/s]
2025-03-19 15:03:59 - Timings: 0.01954 (w/o attn mask) vs 0.01790 (w/ attn mask), ratio: 0.91615, running ratio: 1.11718
  5%|██████                                                                                                                   | 1000/20055 [02:22<46:53,  6.77it/s]
2025-03-19 15:03:59 - Timings: 0.01805 (w/o attn mask) vs 0.01802 (w/ attn mask), ratio: 0.99789, running ratio: 1.11706
```

I would be in favor of providing the attention mask to get rid of the warning, but it would incur a pretty sizable slowdown, so I'm not going to do that.

- Tom Aarsen
